lib/xgu/xgu.h

```diff
@@ -0,0 +1,148 @@
+#include "nv2a_regs.h"
+#pragma GCC diagnostic pop
+
+#include <stdint.h>
+
+typedef enum {
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_UB_D3D     0
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_S1         1
+  XGU_FLOAT = NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_F,
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_UB_OGL     4
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_S32K       5
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_CMP        6
+} XguVertexArrayType;
+
+typedef enum {
+  XGU_END = NV097_SET_BEGIN_END_OP_END, // FIXME: Disallow this one?
+  XGU_POINTS = NV097_SET_BEGIN_END_OP_POINTS,
+  XGU_LINES = NV097_SET_BEGIN_END_OP_LINES,
+  XGU_LINE_LOOP = NV097_SET_BEGIN_END_OP_LINE_LOOP,
+  XGU_LINE_STRIP = NV097_SET_BEGIN_END_OP_LINE_STRIP,
+  XGU_TRIANGLES = NV097_SET_BEGIN_END_OP_TRIANGLES,
+  XGU_TRIANGLE_STRIP = NV097_SET_BEGIN_END_OP_TRIANGLE_STRIP,
+  XGU_TRIANGLE_FAN = NV097_SET_BEGIN_END_OP_TRIANGLE_FAN,
+  XGU_QUADS = NV097_SET_BEGIN_END_OP_QUADS,
+  XGU_QUAD_STRIP = NV097_SET_BEGIN_END_OP_QUAD_STRIP,
+  XGU_POLYGON = NV097_SET_BEGIN_END_OP_POLYGON
+} XguPrimitiveType;
+
+typedef enum {
+  XGU_VERTEX_ARRAY = NV2A_VERTEX_ATTR_POSITION,
+  XGU_NORMAL_ARRAY = NV2A_VERTEX_ATTR_NORMAL,
+  XGU_COLOR_ARRAY = NV2A_VERTEX_ATTR_DIFFUSE,
+  XGU_SECONDARY_COLOR_ARRAY = NV2A_VERTEX_ATTR_SPECULAR,
+  XGU_FOG_ARRAY = NV2A_VERTEX_ATTR_FOG,
+  XGU_TEXCOORD0_ARRAY = NV2A_VERTEX_ATTR_TEXTURE0,
+  XGU_TEXCOORD1_ARRAY = NV2A_VERTEX_ATTR_TEXTURE1,
+  XGU_TEXCOORD2_ARRAY = NV2A_VERTEX_ATTR_TEXTURE2,
+  XGU_TEXCOORD3_ARRAY = NV2A_VERTEX_ATTR_TEXTURE3
+} XguVertexArray;
+
+static inline
+DWORD* push_command(DWORD* p, DWORD command, int nparams) {
+    pb_push(p++, command, nparams);
+    return p;
+}
+
+static inline
+DWORD* push_parameter(DWORD* p, DWORD parameter) {
+    *p++ = parameter;
+    return p;
+}
+
+static inline
+DWORD* push_boolean(DWORD* p, bool enabled) {
+  return push_parameter(p, enabled ? 1 : 0);
+}
+
+static inline
+DWORD* push_command_boolean(DWORD* p, DWORD command, bool enabled) {
+  p = push_command(p, command, 1);
+  p = push_boolean(p, enabled);
+  return p;
+}
+
+static inline
+DWORD* push_command_parameter(DWORD* p, DWORD command, DWORD parameter) {
+  p = push_command(p, command, 1);
+  p = push_parameter(p, parameter);
+  return p;
+}
+
+static inline
+DWORD* xgu_begin(DWORD* p, XguPrimitiveType type) {
+
+  // Force people to use xgu_end instead
+  assert(type != NV097_SET_BEGIN_END_OP_END);
+
+  return push_command_parameter(p, NV097_SET_BEGIN_END, type);
+}
+
+static inline
+DWORD* xgu_end(DWORD* p) {
+  return push_command_parameter(p, NV097_SET_BEGIN_END, NV097_SET_BEGIN_END_OP_END);
+}
+
+inline
+DWORD* xgu_clear_color(DWORD* p, DWORD color) {
+    p = push_command_parameter(p, NV097_SET_COLOR_CLEAR_VALUE, color);
+    return p;
+}
+
+inline
+DWORD* xgu_clear(DWORD* p) {
+    /* Should probably do some magic related to 
```

Comment:
Regardless of what GPU command this function would wrap around: it would almost always wrap around only a single function, so you'd split this into:

- `xgu_set_clear_rect_horizontal` (`NV097_SET_CLEAR_RECT_HORIZONTAL`)
- `xgu_set_zstencil_clear_value` (`NV097_SET_ZSTENCIL_CLEAR_VALUE`)
- `xgu_clear_surface` (`NV097_CLEAR_SURFACE`)

= one function per GPU command. If the GL function is `glFoobar`, then you'd look for any register named "FOOBAR" in nv2a_regs.h [note: only those starting in `NV097_`] and you could assume that they might be necessary and create those couple of `xgu_` functions (it's almost guaranteed we have enough framework for the GL function then). If there's too many registers (or none) for `FOOBAR`, then you can just ask on XboxDev Discord, and we'll assist.